src/goto-symex/goto_symex.cpp

```
@@ -80,8 +86,204 @@ void goto_symext::symex_assign(statet &state, const code_assignt &code)
     if(state.source.pc->source_location.get_hide())
       assignment_type = symex_targett::assignment_typet::HIDDEN;
 
+    symex_assignt symex_assign{
+      state, assignment_type, ns, symex_config, target};
+
+    if(constant_propagate_assignment_with_side_effects(
+         state, symex_assign, lhs, rhs))
+      return;
+
     exprt::operandst lhs_if_then_else_conditions;
-    symex_assignt{state, assignment_type, ns, symex_config, target}.assign_rec(
+    symex_assign.assign_rec(
       lhs, expr_skeletont{}, rhs, lhs_if_then_else_conditions);
   }
 }
+
+bool goto_symext::constant_propagate_assignment_with_side_effects(
```

Suggest a better term than  `assignment_with_side_effects` might be `constraint`? It looks like this is constraining its two operands to have a particular relationship, rather than acting as the pure function the `ID_function_application` suggests. @allredj thoughts?

---

The `lhs` argument is currently not used (it's not needed for the string operations currently handled but might be needed in the future). The `rhs` is a `function_application_exprt` with the first two arguments being output parameters (so it's not a pure a function, a bit unfortunate I know). So overall here we handle the potential side effects of `lhs = rhs` since the evaluation of the `rhs` has side effects.

---

It's a bit weird that these only apply at the root of an assignment though. How about instead recursing over the LHS and RHS with something like `propagate_expression_side_effects`, and if it wanted to also propagate a return value it would just replace the expression with its result and let that be propagated as an ordinary assignment?

---

We could of course recurse into the operands. Since we currently don't need that though, can we leave this for a future PR?

---

>  but might be needed in the future

@danpoe, are there any operations where this is actually used?

---

No, and I think that's ok.